Migrate down between 50 and 49 fails in MySQL (when you have sandboxing configured?)

### Describe the bug

just try to do a migrate down after you configured sandboxing

### To Reproduce

1) start Metabase on 49 and mysql as the app db (I think this is key, I haven't tried with H2)
2) configure sandboxing
3) then migrate down
4) see the error

### Expected behavior

It should migrate down

### Logs

```
2024-06-19 23:59:41,717 INFO db.liquibase :: Rolling back app database schema to version 49
liquibase.exception.CommandExecutionException: liquibase.exception.LiquibaseException: liquibase.exception.RollbackFailedException: liquibase.exception.DatabaseException: (conn=20) Cannot add or update a child row: a foreign key constraint fails (`metabase`.`#sql-1_14`, CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `permissions` (`id`) ON DELETE CASCADE) [Failed SQL: (1452) ALTER TABLE `metabase`.`sandboxes` ADD CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `metabase`.`permissions` (`id`) ON DELETE CASCADE]
	at liquibase.command.CommandScope.execute(CommandScope.java:253)
	at liquibase.Liquibase.lambda$rollback$7(Liquibase.java:579)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.Liquibase.runInScope(Liquibase.java:1419)
	at liquibase.Liquibase.rollback(Liquibase.java:568)
	at liquibase.Liquibase.rollback(Liquibase.java:551)
	at liquibase.Liquibase.rollback(Liquibase.java:542)
	at metabase.db.liquibase$rollback_major_version$fn__44512.invoke(liquibase.clj:485)
	at metabase.db.liquibase$run_in_scope_locked$reify__44426.run(liquibase.clj:324)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at metabase.db.liquibase$run_in_scope_locked.invokeStatic(liquibase.clj:317)
	at metabase.db.liquibase$run_in_scope_locked.invoke(liquibase.clj:300)
	at metabase.db.liquibase$rollback_major_version.invokeStatic(liquibase.clj:478)
	at metabase.db.liquibase$rollback_major_version.invoke(liquibase.clj:465)
	at metabase.db.liquibase$rollback_major_version.invokeStatic(liquibase.clj:470)
	at metabase.db.liquibase$rollback_major_version.invoke(liquibase.clj:465)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:671)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.db.setup$migrate_BANG_$fn__53357.invoke(setup.clj:86)
	at metabase.db.liquibase$do_with_liquibase$f_STAR___44367.invoke(liquibase.clj:139)
	at metabase.db.liquibase$do_with_liquibase.invokeStatic(liquibase.clj:142)
	at metabase.db.liquibase$do_with_liquibase.invoke(liquibase.clj:130)
	at metabase.db.setup$migrate_BANG_.invokeStatic(setup.clj:73)
	at metabase.db.setup$migrate_BANG_.doInvoke(setup.clj:55)
	at clojure.lang.RestFn.invoke(RestFn.java:425)
	at metabase.cmd.migrate$migrate_BANG_.invokeStatic(migrate.clj:8)
	at metabase.cmd.migrate$migrate_BANG_.invoke(migrate.clj:5)
	at clojure.lang.Var.invoke(Var.java:384)
	at metabase.cmd$migrate.invokeStatic(cmd.clj:66)
	at metabase.cmd$migrate.invoke(cmd.clj:62)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.cmd$run_cmd$fn__111564.invoke(cmd.clj:301)
	at metabase.cmd$run_cmd.invokeStatic(cmd.clj:300)
	at metabase.cmd$run_cmd.invoke(cmd.clj:290)
	at clojure.lang.Var.invoke(Var.java:388)
	at metabase.core$run_cmd.invokeStatic(core.clj:192)
	at metabase.core$run_cmd.invoke(core.clj:190)
	at metabase.core$entrypoint.invokeStatic(core.clj:214)
	at metabase.core$entrypoint.doInvoke(core.clj:209)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:705)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:662)
	at metabase.bootstrap$_main.invokeStatic(bootstrap.clj:31)
	at metabase.bootstrap$_main.doInvoke(bootstrap.clj:28)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at metabase.bootstrap.main(Unknown Source)
Caused by: liquibase.exception.LiquibaseException: liquibase.exception.RollbackFailedException: liquibase.exception.DatabaseException: (conn=20) Cannot add or update a child row: a foreign key constraint fails (`metabase`.`#sql-1_14`, CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `permissions` (`id`) ON DELETE CASCADE) [Failed SQL: (1452) ALTER TABLE `metabase`.`sandboxes` ADD CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `metabase`.`permissions` (`id`) ON DELETE CASCADE]
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:151)
	at liquibase.command.core.AbstractRollbackCommandStep.doRollback(AbstractRollbackCommandStep.java:112)
	at liquibase.command.core.AbstractRollbackCommandStep.doRollback(AbstractRollbackCommandStep.java:82)
	at liquibase.command.core.RollbackCountCommandStep.lambda$run$0(RollbackCountCommandStep.java:48)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.command.core.RollbackCountCommandStep.run(RollbackCountCommandStep.java:48)
	at liquibase.command.CommandScope.execute(CommandScope.java:217)
	... 57 more
Caused by: liquibase.exception.RollbackFailedException: liquibase.exception.DatabaseException: (conn=20) Cannot add or update a child row: a foreign key constraint fails (`metabase`.`#sql-1_14`, CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `permissions` (`id`) ON DELETE CASCADE) [Failed SQL: (1452) ALTER TABLE `metabase`.`sandboxes` ADD CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `metabase`.`permissions` (`id`) ON DELETE CASCADE]
	at liquibase.changelog.ChangeSet.rollback(ChangeSet.java:956)
	at liquibase.changelog.visitor.RollbackVisitor.visit(RollbackVisitor.java:63)
	at liquibase.changelog.ChangeLogIterator$2.lambda$run$0(ChangeLogIterator.java:133)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.changelog.ChangeLogIterator$2.run(ChangeLogIterator.java:122)
	at liquibase.Scope.lambda$child$0(Scope.java:186)
	at liquibase.Scope.child(Scope.java:195)
	at liquibase.Scope.child(Scope.java:185)
	at liquibase.Scope.child(Scope.java:164)
	at liquibase.Scope.child(Scope.java:252)
	at liquibase.Scope.child(Scope.java:256)
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:89)
	... 66 more
Caused by: liquibase.exception.DatabaseException: (conn=20) Cannot add or update a child row: a foreign key constraint fails (`metabase`.`#sql-1_14`, CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `permissions` (`id`) ON DELETE CASCADE) [Failed SQL: (1452) ALTER TABLE `metabase`.`sandboxes` ADD CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `metabase`.`permissions` (`id`) ON DELETE CASCADE]
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:470)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:77)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:179)
	at liquibase.database.AbstractJdbcDatabase.execute(AbstractJdbcDatabase.java:1303)
	at liquibase.database.AbstractJdbcDatabase.executeRollbackStatements(AbstractJdbcDatabase.java:1327)
	at liquibase.changelog.ChangeSet.rollback(ChangeSet.java:921)
	... 80 more
Caused by: java.sql.SQLIntegrityConstraintViolationException: (conn=20) Cannot add or update a child row: a foreign key constraint fails (`metabase`.`#sql-1_14`, CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `permissions` (`id`) ON DELETE CASCADE)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.createException(ExceptionFactory.java:70)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.create(ExceptionFactory.java:158)
	at org.mariadb.jdbc.MariaDbStatement.executeExceptionEpilogue(MariaDbStatement.java:262)
	at org.mariadb.jdbc.MariaDbStatement.executeInternal(MariaDbStatement.java:362)
	at org.mariadb.jdbc.MariaDbStatement.execute(MariaDbStatement.java:500)
	at com.mchange.v2.c3p0.impl.NewProxyStatement.execute(NewProxyStatement.java:75)
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:464)
	... 85 more
Caused by: org.mariadb.jdbc.internal.util.exceptions.MariaDbSqlException: Cannot add or update a child row: a foreign key constraint fails (`metabase`.`#sql-1_14`, CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `permissions` (`id`) ON DELETE CASCADE)
	at org.mariadb.jdbc.internal.util.exceptions.MariaDbSqlException.of(MariaDbSqlException.java:34)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.exceptionWithQuery(AbstractQueryProtocol.java:195)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:263)
	at org.mariadb.jdbc.MariaDbStatement.executeInternal(MariaDbStatement.java:356)
	... 88 more
Caused by: java.sql.SQLException: Cannot add or update a child row: a foreign key constraint fails (`metabase`.`#sql-1_14`, CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `permissions` (`id`) ON DELETE CASCADE)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.readErrorPacket(AbstractQueryProtocol.java:1693)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.readPacket(AbstractQueryProtocol.java:1555)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.getResult(AbstractQueryProtocol.java:1518)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:257)
	... 89 more
Command failed with exception: liquibase.exception.LiquibaseException: liquibase.exception.RollbackFailedException: liquibase.exception.DatabaseException: (conn=20) Cannot add or update a child row: a foreign key constraint fails (`metabase`.`#sql-1_14`, CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `permissions` (`id`) ON DELETE CASCADE) [Failed SQL: (1452) ALTER TABLE `metabase`.`sandboxes` ADD CONSTRAINT `fk_sandboxes_ref_permissions` FOREIGN KEY (`permission_id`) REFERENCES `metabase`.`permissions` (`id`) ON DELETE CASCADE]
```

### Information about your Metabase installation

```JSON
v50, but maybe this happens also in previous versions?
```


### Severity

P1

### Additional context

The fact that you can't remove the sandboxing to move back to an older version makes it a P1

--- a/enterprise/backend/test/metabase_enterprise/sandbox/api/gtap_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/sandbox/api/gtap_test.clj
@@ -26,7 +26,6 @@
    :card_id              true
    :table_id             true
    :group_id             true
-   :permission_id        false
    :attribute_remappings {:foo 1}})
 
 (defmacro ^:private with-gtap-cleanup
@@ -248,8 +247,7 @@
                         :table_id             table-id-1
                         :group_id             group-id
                         :card_id              card-id-1
-                        :attribute_remappings {:foo 1}
-                        :permission_id        (mt/malli=? [:maybe :int])}]
+                        :attribute_remappings {:foo 1}}]
                       (:sandboxes result)))
               (is (t2/exists? GroupTableAccessPolicy :table_id table-id-1 :group_id group-id))))
 

--- a/enterprise/backend/test/metabase_enterprise/sandbox/api/permissions_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/sandbox/api/permissions_test.clj
@@ -86,7 +86,6 @@
                                 [:group_id             [:= (u/the-id &group)]]
                                 [:table_id             [:= (u/the-id db-2-table)]]
                                 [:card_id              nil?]
-                                [:permission_id        nil?]
                                 [:attribute_remappings nil?]]]
                               (t2/select GroupTableAccessPolicy
                                          :group_id (u/the-id &group)
@@ -98,7 +97,6 @@
                                 [:group_id             [:= (u/the-id other-group)]]
                                 [:table_id             [:= (mt/id :venues)]]
                                 [:card_id              nil?]
-                                [:permission_id        nil?]
                                 [:attribute_remappings nil?]]]
                               (t2/select GroupTableAccessPolicy :group_id (u/the-id other-group)))))))))))))
 

--- a/resources/migrations/001_update_migrations.yaml
+++ b/resources/migrations/001_update_migrations.yaml
@@ -7894,6 +7894,26 @@ databaseChangeLog:
       rollback: # no rollback necessary here too
 
   - changeSet:
+      id: v50.2024-06-20T13:21:30
+      author: noahmoss
+      comment: 'Drop permission_id column on sandboxes table to fix down migrations'
+      changes:
+        - dropColumn:
+            tableName: sandboxes
+            columnName: permission_id
+      rollback:
+        - addColumn:
+            tableName: sandboxes
+            columns:
+              - column:
+                  name: permission_id
+                  type: int
+                  remarks: 'The ID of the corresponding permissions path for this sandbox (deprecated)'
+                  constraints:
+                    nullable: true
+                  defaultValue: null
+
+  - changeSet:
       id: v51.2024-05-13T15:30:57
       author: metamben
       comment: Backup of dataset_query rewritten by the metrics v2 migration

--- a/test/metabase/db/schema_migrations_test.clj
+++ b/test/metabase/db/schema_migrations_test.clj
@@ -2016,6 +2016,39 @@
         (is (= #{(format "/block/db/%d/" db-id)}
                (t2/select-fn-set :object (t2/table-name :model/Permissions) :group_id group-id)))))))
 
+(deftest sandboxing-rollback-test
+  (testing "Can we rollback to 49 when sandboxing is configured"
+    (impl/test-migrations ["v50.2024-01-10T03:27:29" "v50.2024-06-20T13:21:30"] [migrate!]
+      (let [db-id         (first (t2/insert-returning-pks! (t2/table-name Database) {:name       "DB"
+                                                                                     :engine     "h2"
+                                                                                     :created_at :%now
+                                                                                     :updated_at :%now
+                                                                                     :details    "{}"}))
+            table-id      (first (t2/insert-returning-pks! (t2/table-name Table) {:db_id      db-id}
+                                                                               :schema     "SchemaName"
+                                                                               :name       "Table"
+                                                                               :created_at :%now
+                                                                               :updated_at :%now
+                                                                               :active     true))
+            permission-id (t2/insert-returning-pk! (t2/table-name :model/Permissions)
+                                                   {:object "/db/fake-permission/"
+                                                    :group_id 1})
+            _             (t2/query-one {:insert-into :sandboxes
+                                         :values      [{:group_id             1
+                                                        :table_id             table-id
+                                                        :attribute_remappings "{\"foo\", 1}"
+                                                        :permission_id        permission-id}]})]
+        (migrate!)
+        (is (=? {:group_id             1
+                 :table_id             table-id
+                 :attribute_remappings "{\"foo\", 1}"}
+                (t2/select-one :sandboxes :table_id table-id)))
+        (migrate! :down 49)
+        (is (=? {:group_id             1
+                 :table_id             table-id
+                 :attribute_remappings "{\"foo\", 1}"}
+                (t2/select-one :sandboxes :table_id table-id)))))))
+
 (deftest view-count-test
   (testing "report_card.view_count and report_dashboard.view_count should be populated"
     (impl/test-migrations ["v50.2024-04-25T16:29:31" "v50.2024-04-25T16:29:36"] [migrate!]

--- a/test/metabase/db/schema_migrations_test.clj
+++ b/test/metabase/db/schema_migrations_test.clj
@@ -2024,30 +2024,26 @@
                                                                                      :created_at :%now
                                                                                      :updated_at :%now
                                                                                      :details    "{}"}))
-            table-id      (first (t2/insert-returning-pks! (t2/table-name Table) {:db_id      db-id}
-                                                                               :schema     "SchemaName"
-                                                                               :name       "Table"
-                                                                               :created_at :%now
-                                                                               :updated_at :%now
-                                                                               :active     true))
-            permission-id (t2/insert-returning-pk! (t2/table-name :model/Permissions)
-                                                   {:object "/db/fake-permission/"
-                                                    :group_id 1})
+            table-id      (first (t2/insert-returning-pks! (t2/table-name Table) {:db_id      db-id
+                                                                                  :schema     "SchemaName"
+                                                                                  :name       "Table"
+                                                                                  :created_at :%now
+                                                                                  :updated_at :%now
+                                                                                  :active     true}))
+            permission-id (t2/insert-returning-pk! (t2/table-name :model/Permissions) {:object "/db/fake-permission/"
+                                                                                       :group_id 1})
             _             (t2/query-one {:insert-into :sandboxes
                                          :values      [{:group_id             1
                                                         :table_id             table-id
                                                         :attribute_remappings "{\"foo\", 1}"
-                                                        :permission_id        permission-id}]})]
+                                                        :permission_id        permission-id}]})
+            expected        {:group_id             1
+                             :table_id             table-id
+                             :attribute_remappings "{\"foo\", 1}"}]
         (migrate!)
-        (is (=? {:group_id             1
-                 :table_id             table-id
-                 :attribute_remappings "{\"foo\", 1}"}
-                (t2/select-one :sandboxes :table_id table-id)))
+        (is (=? expected (t2/select-one :sandboxes :table_id table-id)))
         (migrate! :down 49)
-        (is (=? {:group_id             1
-                 :table_id             table-id
-                 :attribute_remappings "{\"foo\", 1}"}
-                (t2/select-one :sandboxes :table_id table-id)))))))
+        (is (=? expected (t2/select-one :sandboxes :table_id table-id)))))))
 
 (deftest view-count-test
   (testing "report_card.view_count and report_dashboard.view_count should be populated"

